MagazineLayout/LayoutCore/ModelState.swift

```diff
@@ -730,6 +736,146 @@ final class ModelState {
+      }
+
+      if currentSectionModels[sectionIndex].backgroundModel != nil {
+        backgroundLocationsForFlattenedIndices[flatennedBackgroundIndex] = ElementLocation(
+          elementIndex: 0,
+          sectionIndex: sectionIndex)
+        flatennedBackgroundIndex += 1
+      }
+
+      for itemIndex in 0..<currentSectionModels[sectionIndex].numberOfItems {
+        itemLocationsForFlattenedIndices[flatennedItemIndex] = ElementLocation(
+          elementIndex: itemIndex,
+          sectionIndex: sectionIndex)
+        flatennedItemIndex += 1
+      }
+    }
+  }
+
+  private func elementLocationFramePairsForElements(
+    in rect: CGRect,
+    withElementLocationsForFlattenedIndices elementLocationsForFlattenedIndices: [Int: ElementLocation],
+    andFramesProvidedBy frameProvider: ((ElementLocation) -> CGRect))
+    -> ElementLocationFramePairs
+  {
+    var elementLocationFramePairs = ElementLocationFramePairs()
+
+    guard
+      let indexOfFirstElement = indexOfFirstElement(
+        in: rect,
+        withElementLocationsForFlattenedIndices: elementLocationsForFlattenedIndices,
+        andFramesProvidedBy: frameProvider) else
+    {
+      return elementLocationFramePairs
+    }
+
+    // Used to handle the case where we encounter an element that doesn't intersect the rect, but
+    // previous elements in the same row might.
+    var minYOfNonIntersectingElement: CGFloat?
+
+    // Look backward to find visible elements
+    for elementLocationIndex in (0..<indexOfFirstElement).reversed() {
+      let elementLocation = self.elementLocation(
+        forFlattenedIndex: elementLocationIndex,
+        in: elementLocationsForFlattenedIndices)
+      let frame = frameProvider(elementLocation)
+
+      guard frame.maxY > rect.minY else {
+        if let minY = minYOfNonIntersectingElement, frame.minY < minY {
+          // We're in a previous row, so we know we've captured all intersecting rects for the
+          // subsequent row.
+          break
+        } else {
+          // We've found a non-intersecting item, but still need to check other items in the same
+          // row.
+          minYOfNonIntersectingElement = frame.minY
+          continue
+        }
+      }
+
+      elementLocationFramePairs.append(
+        ElementLocationFramePair(elementLocation: elementLocation, frame: frame))
+    }
+
+    // Look forward to find visible elements
+    for elementLocationIndex in indexOfFirstElement..<elementLocationsForFlattenedIndices.count {
+      let elementLocation = self.elementLocation(
+        forFlattenedIndex: elementLocationIndex,
+        in: elementLocationsForFlattenedIndices)
+      let frame = frameProvider(elementLocation)
+      guard frame.minY < rect.maxY else { break }
+
+      elementLocationFramePairs.append(
+        ElementLocationFramePair(elementLocation: elementLocation, frame: frame))
+    }
+
+    return elementLocationFramePairs
+  }
+
+  private func indexOfFirstElement(
```

Comment:
This name confused me a bit, is this actually the first element that intersects the rect? Or is it just the first element we find?

Reply:
First element we find - I agree the name should be tweaked to reflect that